File: Mobile/PlatziFlixAndroid/app/src/main/java/com/espaciotiago/platziflixandroid/data/repositories/RemoteRatingRepository.kt
package com.espaciotiago.platziflixandroid.data.repositories

import com.espaciotiago.platziflixandroid.data.entities.RatingRequestDTO
import com.espaciotiago.platziflixandroid.data.mappers.RatingMapper
import com.espaciotiago.platziflixandroid.data.network.ApiService
import com.espaciotiago.platziflixandroid.domain.models.CourseRating
import com.espaciotiago.platziflixandroid.domain.models.RatingStats
import com.espaciotiago.platziflixandroid.domain.repositories.RatingRepository
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext

/**
 * Implementation of RatingRepository that fetches data from remote API
 */
class RemoteRatingRepository(
    private val apiService: ApiService
) : RatingRepository {

    override suspend fun getRatingStats(courseId: Int): Result<RatingStats> {
        return withContext(Dispatchers.IO) {
            try {
                val response = apiService.getRatingStats(courseId)
                if (response.isSuccessful) {
                    val statsDTO = response.body()
                    if (statsDTO != null) {
                        Result.success(RatingMapper.fromStatsDTO(statsDTO))
                    } else {
                        Result.failure(Exception("Empty response body for rating stats"))
                    }
                } else {
                    Result.failure(
                        Exception("Failed to fetch rating stats: ${response.code()} ${response.message()}")
                    )
                }
            } catch (exception: Exception) {
                Result.failure(exception)
            }
        }
    }

    override suspend fun getUserRating(courseId: Int, userId: Int): Result<CourseRating?> {
        return withContext(Dispatchers.IO) {
            try {
                val response = apiService.getUserRating(courseId, userId)
                when {
                    response.code() == 204 -> Result.success(null)
                    response.isSuccessful -> {
                        val ratingDTO = response.body()
                        if (ratingDTO != null) {
                            Result.success(RatingMapper.fromDTO(ratingDTO))
                        } else {
                            Result.success(null)
                        }
                    }
                    else -> Result.failure(
                        Exception("Failed to fetch user rating: ${response.code()} ${response.message()}")
                    )
                }
            } catch (exception: Exception) {
                Result.failure(exception)
            }
        }
    }

    override suspend fun upsertRating(courseId: Int, userId: Int, stars: Int): Result<CourseRating> {
        return withContext(Dispatchers.IO) {
            try {
                val response = apiService.upsertRating(
                    courseId = courseId,
                    request = RatingRequestDTO(userId = userId, rating = stars)
                )
                if (response.code() == 201) {
                    val ratingDTO = response.body()
                    if (ratingDTO != null) {
                        Result.success(RatingMapper.fromDTO(ratingDTO))
                    } else {
                        Result.failure(Exception("Empty response body for upsert rating"))
                    }
                } else {
                    Result.failure(
                        Exception("Failed to upsert rating: ${response.code()} ${response.message()}")
                    )
                }
            } catch (exception: Exception) {
                Result.failure(exception)
            }
        }
    }
}

File: Mobile/PlatziFlixiOS/PlatziFlixiOS/Data/Repositories/RemoteRatingRepository.swift
import Foundation

/// Remote implementation of RatingRepository
/// Fetches and persists course rating data using NetworkService
final class RemoteRatingRepository: RatingRepository {

    // MARK: - Properties
    private let networkService: NetworkService

    // MARK: - Initialization
    init(networkService: NetworkService = NetworkManager.shared) {
        self.networkService = networkService
    }

    // MARK: - RatingRepository Implementation

    /// Fetches the aggregated rating statistics for a course.
    /// `average_rating == 0.0` is a valid response when the course has no ratings.
    func getRatingStats(courseId: Int) async throws -> RatingStats {
        let endpoint = RatingAPIEndpoints.getRatingStats(courseId: courseId)
        let dto = try await networkService.request(endpoint, responseType: RatingStatsDTO.self)
        return RatingMapper.toDomain(dto)
    }

    /// Fetches the rating a user has given to a course.
    /// The backend responds `204 No Content` (empty body) when the user has not rated yet.
    func getUserRating(courseId: Int, userId: Int) async throws -> CourseRating? {
        let endpoint = RatingAPIEndpoints.getUserRating(courseId: courseId, userId: userId)
        let data = try await networkService.request(endpoint)

        guard !data.isEmpty else { return nil }

        do {
            let decoder = JSONDecoder()
            decoder.dateDecodingStrategy = .iso8601
            let dto = try decoder.decode(RatingResponseDTO.self, from: data)
            return RatingMapper.toDomain(dto)
        } catch {
            throw NetworkError.decodingError(error)
        }
    }

    /// Creates or updates (upsert) a user's rating for a course.
    /// The backend always responds `201 Created`, even when updating an existing rating.
    func upsertRating(courseId: Int, userId: Int, stars: Int) async throws -> CourseRating {
        let endpoint = RatingAPIEndpoints.upsertRating(courseId: courseId, userId: userId, rating: stars)
        let dto = try await networkService.request(endpoint, responseType: RatingResponseDTO.self)
        return RatingMapper.toDomain(dto)
    }
}
